dulwich/objects.py

```diff
@@ -383,8 +384,23 @@ def _serialize(self):
         raise NotImplementedError(self._serialize)
 
     @classmethod
+    def _check_is_corrupted(cls, path):
```

Comment:
for performance reasons, please raise exceptions about formatting issues where they are noticed, rather than checking for all corruption up front. E.g. in ShaFile._parse_file, raise an exception if the file was empty.